source/common/http/utility.cc

```diff
@@ -387,6 +388,115 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
+  for (const auto& token : connection_header_tokens) {
+
+    const absl::string_view token_sv = StringUtil::trim(token);
+
+    // If the Connection token value is not a nominated header, ignore it here since
+    // the connection header is removed elsewhere when the H1 request is upgraded to H2
+    if (StringUtil::CaseInsensitiveCompare()(token_sv, cv.Close) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.Http2Settings) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.KeepAlive) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.Upgrade)) {
+      continue;
+    }
+
+    // Build the LowerCaseString for header lookup
+    const std::string header_to_remove = std::string(token_sv);
+    const LowerCaseString lcs_header_to_remove(header_to_remove);
+
+    // By default we will remove any nominated headers
+    bool keep_header = false;
+
+    // Determine whether the nominated header contains invalid values
+    HeaderEntry* nominated_header = NULL;
+
+    if (StringUtil::CaseInsensitiveCompare()(token_sv, Http::Headers::get().Connection.get())) {
+      // Remove the connection header from the nominated tokens if it's self nominated
+      // The connection header itself is *not removed*
+      ENVOY_LOG_MISC(trace, "Skipping self nominated header [{}]", token_sv);
+      keep_header = true;
+      headers_to_remove.emplace(token_sv);
+
+    } else if (StringUtil::CaseInsensitiveCompare()(token_sv,
+                                                    Http::Headers::get().ForwardedFor.get()) ||
+               StringUtil::CaseInsensitiveCompare()(token_sv,
+                                                    Http::Headers::get().ForwardedHost.get()) ||
+               StringUtil::CaseInsensitiveCompare()(token_sv,
+                                                    Http::Headers::get().ForwardedProto.get()) ||
+               !token_sv.find(':')) {
+      // If pseudo headers or X-Forwarded* headers are nominated, this could be
+      // an invalid request. Reject the request
+      return false;
+
+    } else {
+      // Examine the value of all other nominated headers
+      nominated_header = headers.get(lcs_header_to_remove);
+    }
+
+    if (nominated_header) {
+      const HeaderString& nominated_header_value = nominated_header->value();
+      auto nominated_heaader_value_sv = nominated_header_value.getStringView();
+
+      const bool is_te_header =
+          StringUtil::CaseInsensitiveCompare()(token_sv, Http::Headers::get().TE.get());
+
+      // reject the request if the TE header is too large
+      if (is_te_header && (nominated_heaader_value_sv.size() >= 256)) {
+        return false;
+      }
+
+      // If the TE header contains "trailers" we will reset the header to this value since it's
+      // the only permitted value
+      if (is_te_header && (nominated_heaader_value_sv.find(
+                               Http::Headers::get().TEValues.Trailers) != std::string::npos)) {
```

Comment:
I'll defer to other reviewers to see this check is enough. It does not account for a weird edge case where a header name contains the "trailers" keyword. I thought about it more and I think your original version using StrSplit in the for range loop was correct. The StrSplit in the for range loop does not cause memory allocation and with the length check it would be safe. If the find check is deemed to be insufficient you should use check for the trailers keyword using your original method. Sorry abot flip-flopping on this one.

Reply:
This check should only execute if the header itself is "TE".   I removed the tokenization since it was expensive (per a comment on the old PR).  In this change, if the header is "TE" and we find "trailers" and other values (there should be a check for whether a comma exists in the header value), then we would replace the existing value of `TE: whatever, and, trailers` with just `TE: trailers`.

There was a second commit where I fixed the variable spelling error and added this check.  In any case, I'm open to what others would suggest here.